lib/Conversion/LLHDToLLVM/LLHDToLLVM.cpp

```diff
@@ -1336,69 +2159,411 @@ struct ExtractSliceOpConversion : public ConvertToLLVMPattern {
+
+        auto adjusted = shiftIntegerSigPointer(
+            op->getLoc(), &getDialect(), rewriter, sigDetail[0], adjustedStart);
+        // Create a new subsignal with the new pointer and offset.
+        rewriter.replaceOp(op, createSubSig(&getDialect(), rewriter,
+                                            op->getLoc(), sigDetail,
+                                            adjusted.first, adjusted.second));
+      } else if (auto arrTy = resTy.getUnderlyingType().dyn_cast<ArrayType>()) {
+        auto llvmArrTy =
+            typeConverter.convertType(arrTy).cast<LLVM::LLVMType>();
+
+        auto adjustedPtr =
+            shiftArraySigPointer(op->getLoc(), rewriter, llvmArrTy,
+                                 sigDetail[0], transformed.start());
+        rewriter.replaceOp(op,
+                           createSubSig(&getDialect(), rewriter, op->getLoc(),
+                                        sigDetail, adjustedPtr, sigDetail[1]));
+      }
+      return success();
+    }
+
+    if (auto arrTy = extsOp.result().getType().dyn_cast<ArrayType>()) {
+      auto elemTy = typeConverter.convertType(arrTy.getElementType())
+                        .cast<LLVM::LLVMType>();
+      auto llvmArrTy = typeConverter.convertType(arrTy).cast<LLVM::LLVMType>();
+      auto targetTy = transformed.target().getType().cast<LLVM::LLVMType>();
+
+      auto zeroC = rewriter.create<LLVM::ConstantOp>(
+          op->getLoc(), i64Ty, rewriter.getI64IntegerAttr(0));
+      auto oneC = rewriter.create<LLVM::ConstantOp>(
+          op->getLoc(), i64Ty, rewriter.getI64IntegerAttr(1));
+      auto zextStart = zextByOne(op->getLoc(), rewriter, transformed.start());
+
+      // LLVM::ExtractValueOp only takes attribute arguments for the indexes, so
+      // we need to store the array into the stack and use gep+load to get the
+      // elements dynamically.
+      auto targetPtr = rewriter.create<LLVM::AllocaOp>(
+          op->getLoc(), targetTy.getPointerTo(), ArrayRef<Value>(oneC));
+      rewriter.create<LLVM::StoreOp>(op->getLoc(), transformed.target(),
+                                     targetPtr);
+      Value slice = rewriter.create<LLVM::UndefOp>(op->getLoc(), llvmArrTy);
+      for (size_t i = 0, e = arrTy.getLength(); i < e; ++i) {
+        auto indexC = rewriter.create<LLVM::ConstantOp>(
+            op->getLoc(), zextStart.getType(), rewriter.getI64IntegerAttr(i));
+        auto adjustedIndex =
+            rewriter.create<LLVM::AddOp>(op->getLoc(), indexC, zextStart);
+        auto gep = rewriter.create<LLVM::GEPOp>(
+            op->getLoc(), elemTy.getPointerTo(), targetPtr,
+            ArrayRef<Value>({zeroC, adjustedIndex}));
+        auto extract = arrayBoundaryCheck(op->getLoc(), rewriter, llvmArrTy,
+                                          targetPtr, gep);
+        slice = rewriter.create<LLVM::InsertValueOp>(
+            op->getLoc(), llvmArrTy, slice, extract,
+            rewriter.getI32ArrayAttr(i));
       }
```

Comment:
This looks so much cleaner now that it doesn't need to be ham-fisted into the shift operation. Super nice!